### DISCUSSION TITLE
Documenting/Notes on overload resolution behaviour

### DISCUSSION THREAD

**Discussion by realvictorprm** [dotnet/fsharp#7218] [General]
After a talk with Phillip Carter I'll try to document the current overload resolution behaviour here. People with deep knowledge about this (especially you @dsyme!) comment if you can add information to unclear parts & also please comment if specific parts of this are wrong.

# Open questions and notes about overload resolution

Starting at `14.5.4 Solving Member Constraints`:
```
[...] A member constraint is satisfied if one of the types in the support set type1 ... typen satisfies the
member constraint. [...]
```

This leaves interpretation open for the implementation if multiple overloads would be a fit for the current constraint (did not find a section about how multiple fitting overloads should be handled).

----
## Behaviour on multiple fitting overloads:

To my understanding of the code in the compiler, behaviour with multiple overloads is following.

1. If there are multiple overloads fitting the current constraint signature failwith an overload resolution error if the support contains only not `free` / specific resolved types.
2. If there are multiple overloads fitting the current constraint signature without the complete support consisting of specific resolved types do not fail & delay overload resolution, whatever this means in detail.

The delayed overload resolution isn't well enough documented for me to get a full crasp of what is expected to happen (which may be a reason why there are bugs in there afterall).

This hole in specification and documentation allows one to abuse the current implemented mechanism in such ways that:
1. 2..N methods with equal name, equal argument count but different type signature can be invoked via a single method exploiting the delayed overload resolution via including one of the parameters or return type in the support. This effectively delays overload resolution until the full type signature is available.

**Comment by gusty** [comment]
It shouldn't be that complicated to get a detailed spec.

To me it sounds like delayed resolution should happen when there are 2 or more candidates with the same "score".

This means, that overload resolution ambiguity errors should never occur (unless the function is not inline). If there is ambiguity => delay it.
